sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -452,26 +490,62 @@ class SpillableDeviceBufferHandle private (
     materialized
   }
 
+  private var toSpill: Option[DeviceMemoryBuffer] = None
+  private var spilled: Option[DeviceMemoryBuffer] = None
+  override def releaseDeviceResource(): Unit = {
+    synchronized {
+      spilled.foreach(_.close())
+      spilled = None
+    }
+    super.releaseDeviceResource()
+  }
+
   override def spill(): Long = {
     if (!spillable) {
       0L
     } else {
-      synchronized {
-        if (host.isEmpty && dev.isDefined) {
-          host = Some(SpillableHostBufferHandle.createHostHandleFromDeviceBuff(dev.get))
-          sizeInBytes
+      val thisThreadSpills = synchronized {
+        if (host.isEmpty && dev.isDefined && toSpill.isEmpty) {
+          toSpill = dev
+          // incRefCount here so that if close() is called
+          // while we are spilling, we will prevent the buffer being freed
+          toSpill.get.incRefCount()
+          true
         } else {
-          0L
+          false
         }
       }
+      if (thisThreadSpills) {
+        val buf = toSpill.get
+        withResource(buf) { _ =>
+          // the spill IO is non-blocking as it won't impact dev or host directly
+          // instead we "atomically" swap the buffers below once they are ready
+          val stagingHost =
+            Some(SpillableHostBufferHandle.createHostHandleFromDeviceBuff(buf))
+          synchronized {
+            if (!closed) {
+              host = stagingHost
+            }
+            spilled = dev
+            dev = None
+            toSpill = None
+          }
+        }
+        sizeInBytes
+      } else {
+        0
+      }
     }
   }
 
   override def close(): Unit = {
+    // do we need to Cuda.deviceSynchronize here?
+    // what if we don't spill
```

Comment:
I don't see a need here. `close` is called from the user code and that's who owns the `handle`, so the caller needs to be careful about calling synchronize before closing memory, but at the same time we use stream aware allocators, so I don't see cases where we need to add extra synchronization.

Reply:
Oh sorry, I know we just discussed that offline, I forgot to delete this comment